osm2streets/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,205 @@
         }
     }
 
-    for (id, complexity) in changes.into_iter() {
-        streets.intersections.get_mut(&id).unwrap().complexity = complexity;
+    for (id, (complexity, conflict_level, movements)) in changes {
+        let intersection = streets.intersections.get_mut(&id).unwrap();
+        intersection.complexity = complexity;
+        intersection.conflict_level = conflict_level;
+        intersection.movements = movements;
     }
 }
 
 /// Guesses the complexity of the intersection based on the connecting roads and their lanes.
 ///
 /// The existing complexity field is ignored, so be careful how you use the guessed value.
-fn guess_complexity(streets: &StreetNetwork, intersection_id: &NodeID) -> IntersectionComplexity {
-    let roads = streets.roads_per_intersection(*intersection_id);
+fn guess_complexity(
+    streets: &StreetNetwork,
+    intersection_id: &NodeID,
+) -> (IntersectionComplexity, ConflictType, Vec<IndexedMovement>) {
+    use ConflictType::*;
+    let road_ids = streets.roads_per_intersection(*intersection_id);
+    let roads: Vec<&Road> = road_ids.iter().map(|id| &streets.roads[id]).collect();
 
     // A terminus is characterised by a single connected road.
-    if roads.len() == 1 {
-        return Terminus;
+    if road_ids.len() == 1 {
+        return (Terminus, Uncontested, vec![(0, 0)]);
     }
 
     // A Connection is characterised by exactly two connected roads.
-    if roads.len() == 2 {
-        return Connection;
+    if road_ids.len() == 2 {
+        // TODO check directions of roads and determine movements and if it is well formed etc.
+        return (Connection, Uncontested, vec![(0, 1), (1, 0)]);
     }
 
-    // A MultiConnection is characterised by exactly one dividing line traveling through it (the
-    // line that separates traffic in different directions) and no traffic that crosses it.
-    if roads.len() == 3 {
-        let mut _num_roads_in = 0;
-        let mut _num_roads_out = 0;
-        let mut num_roads_inout = 0;
-        for road_id in roads {
-            let is_outward = road_id.i1 == *intersection_id;
-            let road = streets.roads.get(&road_id).unwrap();
-            match road.oneway_for_driving() {
-                Some(Direction::Fwd) => {
-                    if is_outward {
-                        _num_roads_out += 1
-                    } else {
-                        _num_roads_in += 1
-                    }
+    // Calculate all the possible movements, (except U-turns, for now).
+    let mut connections = Vec::new();
+    // Consider all pairs of roads, from s to d. Identify them using their index in the list - which
+    // is sorted in clockwise order - so that we can compare their position later.
+    for s in 0..road_ids.len() {
+        for d in 0..road_ids.len() {
+            if s == d {
+                continue; // Ignore U-turns.
+            }
+
+            // Calculate if it is possible to emerge from s into the intersection.
```

Comment:
Feels related to #1 and the idea of road halves. In A/B Street, we have a `DirectedRoadID`, which is a road ID and a forwards/backwards direction. I wonder if that'd be useful here too? I started thinking through a method for a road to see which directions point incoming or outgoing to an intersection and return the results in a clockwise order, but... didn't come up with anything clear

Reply:
`DirectedRoadID` is a concept that would be useful here, I agree. It feels like it would be more ergonomic if an intersection stored `DirectedRoadID` for all connected roads in a consistent direction, so that we don't have to do the `.i1 == id` dance.

I wonder if it would be even more ergonomic to have a type that is a "view" of a road that supports inverting the direction. That is, if I want to get the list of roads connected to an intersection, it would be nice to treat them all as if they were all incoming roads that "end" at the intersection. The idea is to "hoist" one of the boolean attributes up out of the scope, so the algorithm has half as many cases to contend with: all roads are incoming, so I don't have to branch every time I want to get the "end of the road that meets the intersection" or anything like that.

For example, it would be cool if I could do something like the psuedo code that follows. `roads_clockwise` would return an iterator of `RoadView` instances that - throughout its API - reverses the meaning of "start"/"end" and "forward"/"backward" relative to the underlying `Road`.

```rust
for road in intersection.roads_clockwise() {
  print("road {} meets the intersection at {}", road.directed_id, road.get_end_point());
}
```

---

The "half road" idea leads me to a different API that would be even more ergonomic in this case. A `HalfRoad` could be a view into an underlying `Road` that is always oriented in the direction of travel.

```rust
for (s, half_road_s) in intersection.half_roads_clockwise().enumerate() {
  if let Incoming(src) = half_road_s {
    for (d, half_road_d) in intersection.half_roads_clockwise().enumerate() {
      if let Outgoing(dst) = half_road_d {
        if turn_is_allowed(src, dst) {
          movements.add(s, d);
        }
      }
    }
  }
}
```

The logic for determining crossing/merging/diverging becomes a lot simpler too, because every position around the intersection boundary (`s` and `d`) is just one direction, so we wouldn't have to factor in the driving side. (The driving side is the boolean attribute that has been hoisted in this case).

--- 

I suspect we will need to discuss this synchronously to get to a full understanding...

Reply:
> That is, if I want to get the list of roads connected to an intersection, it would be nice to treat them all as if they were all incoming roads that "end" at the intersection. 

The intersection geometry algorithm does exactly that too! It's a really useful process, so it definitely ought to be easier.

> A HalfRoad could be a view into an underlying Road that is always oriented in the direction of travel.

I'm really liking this idea! Depending what fields/methods this view has, maybe this is straightforward to implement, or maybe we attempt something with lifetimes. (A `HalfRoad` internally stores a pointer to a `RawRoad`, so we write out explicit lifetimes to keep everything safe.

Happy to chat about this soon! I have plenty of time this week